// vso-node-api/interfaces/TestInterfaces.js
/*
* ---------------------------------------------------------
* Copyright(C) Microsoft Corporation. All rights reserved.
* ---------------------------------------------------------
*
* ---------------------------------------------------------
* Generated file, DO NOT EDIT
* ---------------------------------------------------------
*/
"use strict";
var TfsCoreInterfaces = require("../interfaces/CoreInterfaces");
var VSSInterfaces = require("../interfaces/common/VSSInterfaces");
(function (AttachmentType) {
    AttachmentType[AttachmentType["GeneralAttachment"] = 0] = "GeneralAttachment";
    AttachmentType[AttachmentType["AfnStrip"] = 1] = "AfnStrip";
    AttachmentType[AttachmentType["BugFilingData"] = 2] = "BugFilingData";
    AttachmentType[AttachmentType["CodeCoverage"] = 3] = "CodeCoverage";
    AttachmentType[AttachmentType["IntermediateCollectorData"] = 4] = "IntermediateCollectorData";
    AttachmentType[AttachmentType["RunConfig"] = 5] = "RunConfig";
    AttachmentType[AttachmentType["TestImpactDetails"] = 6] = "TestImpactDetails";
    AttachmentType[AttachmentType["TmiTestRunDeploymentFiles"] = 7] = "TmiTestRunDeploymentFiles";
    AttachmentType[AttachmentType["TmiTestRunReverseDeploymentFiles"] = 8] = "TmiTestRunReverseDeploymentFiles";
    AttachmentType[AttachmentType["TmiTestResultDetail"] = 9] = "TmiTestResultDetail";
    AttachmentType[AttachmentType["TmiTestRunSummary"] = 10] = "TmiTestRunSummary";
})(exports.AttachmentType || (exports.AttachmentType = {}));
var AttachmentType = exports.AttachmentType;
(function (CloneOperationState) {
    CloneOperationState[CloneOperationState["Failed"] = 2] = "Failed";
    CloneOperationState[CloneOperationState["InProgress"] = 1] = "InProgress";
    CloneOperationState[CloneOperationState["Queued"] = 0] = "Queued";
    CloneOperationState[CloneOperationState["Succeeded"] = 3] = "Succeeded";
})(exports.CloneOperationState || (exports.CloneOperationState = {}));
var CloneOperationState = exports.CloneOperationState;
(function (CoverageQueryFlags) {
    /**
     * If set, the Coverage.Modules property will be populated.
     */
    CoverageQueryFlags[CoverageQueryFlags["Modules"] = 1] = "Modules";
    /**
     * If set, the ModuleCoverage.Functions properties will be populated.
     */
    CoverageQueryFlags[CoverageQueryFlags["Functions"] = 2] = "Functions";
    /**
     * If set, the ModuleCoverage.CoverageData field will be populated.
     */
    CoverageQueryFlags[CoverageQueryFlags["BlockData"] = 4] = "BlockData";
})(exports.CoverageQueryFlags || (exports.CoverageQueryFlags = {}));
var CoverageQueryFlags = exports.CoverageQueryFlags;
(function (CustomTestFieldScope) {
    CustomTestFieldScope[CustomTestFieldScope["None"] = 0] = "None";
    CustomTestFieldScope[CustomTestFieldScope["TestRun"] = 1] = "TestRun";
    CustomTestFieldScope[CustomTestFieldScope["TestResult"] = 2] = "TestResult";
    CustomTestFieldScope[CustomTestFieldScope["System"] = 4] = "System";
    CustomTestFieldScope[CustomTestFieldScope["All"] = 7] = "All";
})(exports.CustomTestFieldScope || (exports.CustomTestFieldScope = {}));
var CustomTestFieldScope = exports.CustomTestFieldScope;
(function (CustomTestFieldType) {
    CustomTestFieldType[CustomTestFieldType["Bit"] = 2] = "Bit";
    CustomTestFieldType[CustomTestFieldType["DateTime"] = 4] = "DateTime";
    CustomTestFieldType[CustomTestFieldType["Int"] = 8] = "Int";
    CustomTestFieldType[CustomTestFieldType["Float"] = 6] = "Float";
    CustomTestFieldType[CustomTestFieldType["String"] = 12] = "String";
    CustomTestFieldType[CustomTestFieldType["Guid"] = 14] = "Guid";
})(exports.CustomTestFieldType || (exports.CustomTestFieldType = {}));
var CustomTestFieldType = exports.CustomTestFieldType;
(function (GroupTestResultsBy) {
    GroupTestResultsBy[GroupTestResultsBy["None"] = 0] = "None";
    GroupTestResultsBy[GroupTestResultsBy["AutomatedTestStorage"] = 1] = "AutomatedTestStorage";
})(exports.GroupTestResultsBy || (exports.GroupTestResultsBy = {}));
var GroupTestResultsBy = exports.GroupTestResultsBy;
(function (ResultDetails) {
    ResultDetails[ResultDetails["None"] = 0] = "None";
    ResultDetails[ResultDetails["Iterations"] = 1] = "Iterations";
    ResultDetails[ResultDetails["WorkItems"] = 2] = "WorkItems";
})(exports.ResultDetails || (exports.ResultDetails = {}));
var ResultDetails = exports.ResultDetails;
(function (ResultObjectType) {
    ResultObjectType[ResultObjectType["TestSuite"] = 0] = "TestSuite";
    ResultObjectType[ResultObjectType["TestPlan"] = 1] = "TestPlan";
})(exports.ResultObjectType || (exports.ResultObjectType = {}));
var ResultObjectType = exports.ResultObjectType;
(function (ResultOutcome) {
    ResultOutcome[ResultOutcome["Pass"] = 1] = "Pass";
    ResultOutcome[ResultOutcome["Fail"] = 2] = "Fail";
    ResultOutcome[ResultOutcome["Pending"] = 3] = "Pending";
})(exports.ResultOutcome || (exports.ResultOutcome = {}));
var ResultOutcome = exports.ResultOutcome;
(function (TestConfigurationState) {
    /**
     * The configuration can be used for new test runs.
     */
    TestConfigurationState[TestConfigurationState["Active"] = 1] = "Active";
    /**
     * The configuration has been retired and should not be used for new test runs.
     */
    TestConfigurationState[TestConfigurationState["Inactive"] = 2] = "Inactive";
})(exports.TestConfigurationState || (exports.TestConfigurationState = {}));
var TestConfigurationState = exports.TestConfigurationState;
(function (TestOutcome) {
    /**
     * Only used during an update to preserve the existing value.
     */
    TestOutcome[TestOutcome["Unspecified"] = 0] = "Unspecified";
    /**
     * Test has not been completed, or the test type does not report pass/failure.
     */
    TestOutcome[TestOutcome["None"] = 1] = "None";
    /**
     * Test was executed w/o any issues.
     */
    TestOutcome[TestOutcome["Passed"] = 2] = "Passed";
    /**
     * Test was executed, but there were issues. Issues may involve exceptions or failed assertions.
     */
    TestOutcome[TestOutcome["Failed"] = 3] = "Failed";
    /**
     * Test has completed, but we can't say if it passed or failed. May be used for aborted tests...
     */
    TestOutcome[TestOutcome["Inconclusive"] = 4] = "Inconclusive";
    /**
     * The test timed out
     */
    TestOutcome[TestOutcome["Timeout"] = 5] = "Timeout";
    /**
     * Test was aborted. This was not caused by a user gesture, but rather by a framework decision.
     */
    TestOutcome[TestOutcome["Aborted"] = 6] = "Aborted";
    /**
     * Test had it chance for been executed but was not, as ITestElement.IsRunnable == false.
     */
    TestOutcome[TestOutcome["Blocked"] = 7] = "Blocked";
    /**
     * Test was not executed. This was caused by a user gesture - e.g. user hit stop button.
     */
    TestOutcome[TestOutcome["NotExecuted"] = 8] = "NotExecuted";
    /**
     * To be used by Run level results. This is not a failure.
     */
    TestOutcome[TestOutcome["Warning"] = 9] = "Warning";
    /**
     * There was a system error while we were trying to execute a test.
     */
    TestOutcome[TestOutcome["Error"] = 10] = "Error";
    /**
     * Test is Not Applicable for execution.
     */
    TestOutcome[TestOutcome["NotApplicable"] = 11] = "NotApplicable";
    /**
     * Test is paused.
     */
    TestOutcome[TestOutcome["Paused"] = 12] = "Paused";
    /**
     * Test is currently executing. Added this for TCM charts
     */
    TestOutcome[TestOutcome["InProgress"] = 13] = "InProgress";
    TestOutcome[TestOutcome["MaxValue"] = 13] = "MaxValue";
})(exports.TestOutcome || (exports.TestOutcome = {}));
var TestOutcome = exports.TestOutcome;
(function (TestResultsContextType) {
    TestResultsContextType[TestResultsContextType["Build"] = 1] = "Build";
    TestResultsContextType[TestResultsContextType["Release"] = 2] = "Release";
})(exports.TestResultsContextType || (exports.TestResultsContextType = {}));
var TestResultsContextType = exports.TestResultsContextType;
(function (TestRunState) {
    /**
     * Only used during an update to preserve the existing value.
     */
    TestRunState[TestRunState["Unspecified"] = 0] = "Unspecified";
    /**
     * The run is still being created.  No tests have started yet.
     */
    TestRunState[TestRunState["NotStarted"] = 1] = "NotStarted";
    /**
     * Tests are running.
     */
    TestRunState[TestRunState["InProgress"] = 2] = "InProgress";
    /**
     * All tests have completed or been skipped.
     */
    TestRunState[TestRunState["Completed"] = 3] = "Completed";
    /**
     * Run is stopped and remaing tests have been aborted
     */
    TestRunState[TestRunState["Aborted"] = 4] = "Aborted";
    /**
     * Run is currently initializing This is a legacy state and should not be used any more
     */
    TestRunState[TestRunState["Waiting"] = 5] = "Waiting";
    /**
     * Run requires investigation because of a test point failure This is a legacy state and should not be used any more
     */
    TestRunState[TestRunState["NeedsInvestigation"] = 6] = "NeedsInvestigation";
})(exports.TestRunState || (exports.TestRunState = {}));
var TestRunState = exports.TestRunState;
(function (TestRunSubstate) {
    TestRunSubstate[TestRunSubstate["None"] = 0] = "None";
    TestRunSubstate[TestRunSubstate["CreatingEnvironment"] = 1] = "CreatingEnvironment";
    TestRunSubstate[TestRunSubstate["RunningTests"] = 2] = "RunningTests";
    TestRunSubstate[TestRunSubstate["CanceledByUser"] = 3] = "CanceledByUser";
    TestRunSubstate[TestRunSubstate["AbortedBySystem"] = 4] = "AbortedBySystem";
    TestRunSubstate[TestRunSubstate["TimedOut"] = 5] = "TimedOut";
    TestRunSubstate[TestRunSubstate["PendingAnalysis"] = 6] = "PendingAnalysis";
    TestRunSubstate[TestRunSubstate["Analyzed"] = 7] = "Analyzed";
    TestRunSubstate[TestRunSubstate["CancellationInProgress"] = 8] = "CancellationInProgress";
})(exports.TestRunSubstate || (exports.TestRunSubstate = {}));
var TestRunSubstate = exports.TestRunSubstate;
exports.TypeInfo = {
    AggregatedDataForResultTrend: {
        fields: null
    },
    AggregatedResultsAnalysis: {
        fields: null
    },
    AggregatedResultsByOutcome: {
        fields: null
    },
    AggregatedResultsDifference: {
        fields: null
    },
    AttachmentType: {
        enumValues: {
            "generalAttachment": 0,
            "afnStrip": 1,
            "bugFilingData": 2,
            "codeCoverage": 3,
            "intermediateCollectorData": 4,
            "runConfig": 5,
            "testImpactDetails": 6,
            "tmiTestRunDeploymentFiles": 7,
            "tmiTestRunReverseDeploymentFiles": 8,
            "tmiTestResultDetail": 9,
            "tmiTestRunSummary": 10,
        }
    },
    BatchResponse: {
        fields: null
    },
    BuildConfiguration: {
        fields: null
    },
    BuildCoverage: {
        fields: null
    },
    BuildReference: {
        fields: null
    },
    CloneOperationInformation: {
        fields: null
    },
    CloneOperationState: {
        enumValues: {
            "failed": 2,
            "inProgress": 1,
            "queued": 0,
            "succeeded": 3,
        }
    },
    CloneOptions: {
        fields: null
    },
    CloneStatistics: {
        fields: null
    },
    CodeCoverageData: {
        fields: null
    },
    CodeCoverageStatistics: {
        fields: null
    },
    CodeCoverageSummary: {
        fields: null
    },
    CoverageQueryFlags: {
        enumValues: {
            "modules": 1,
            "functions": 2,
            "blockData": 4,
        }
    },
    CoverageStatistics: {
        fields: null
    },
    CustomTestField: {
        fields: null
    },
    CustomTestFieldDefinition: {
        fields: null
    },
    CustomTestFieldScope: {
        enumValues: {
            "none": 0,
            "testRun": 1,
            "testResult": 2,
            "system": 4,
            "all": 7,
        }
    },
    CustomTestFieldType: {
        enumValues: {
            "bit": 2,
            "dateTime": 4,
            "int": 8,
            "float": 6,
            "string": 12,
            "guid": 14,
        }
    },
    DtlEnvironmentDetails: {
        fields: null
    },
    FailingSince: {
        fields: null
    },
    FunctionCoverage: {
        fields: null
    },
    GroupTestResultsBy: {
        enumValues: {
            "none": 0,
            "automatedTestStorage": 1,
        }
    },
    LastResultDetails: {
        fields: null
    },
    ModuleCoverage: {
        fields: null
    },
    NameValuePair: {
        fields: null
    },
    PlanUpdateModel: {
        fields: null
    },
    PointAssignment: {
        fields: null
    },
    PointUpdateModel: {
        fields: null
    },
    PointWorkItemProperty: {
        fields: null
    },
    QueryModel: {
        fields: null
    },
    ReleaseReference: {
        fields: null
    },
    Response: {
        fields: null
    },
    ResultDetails: {
        enumValues: {
            "none": 0,
            "iterations": 1,
            "workItems": 2,
        }
    },
    ResultObjectType: {
        enumValues: {
            "testSuite": 0,
            "testPlan": 1,
        }
    },
    ResultOutcome: {
        enumValues: {
            "pass": 1,
            "fail": 2,
            "pending": 3,
        }
    },
    ResultRetentionSettings: {
        fields: null
    },
    ResultUpdateRequestModel: {
        fields: null
    },
    ResultUpdateResponseModel: {
        fields: null
    },
    RunCreateModel: {
        fields: null
    },
    RunFilter: {
        fields: null
    },
    RunStatistic: {
        fields: null
    },
    RunUpdateModel: {
        fields: null
    },
    ShallowReference: {
        fields: null
    },
    SharedStepModel: {
        fields: null
    },
    SuiteCreateModel: {
        fields: null
    },
    SuiteTestCase: {
        fields: null
    },
    SuiteUpdateModel: {
        fields: null
    },
    TestActionResultModel: {
        fields: null
    },
    TestAttachmentReference: {
        fields: null
    },
    TestAttachmentRequestModel: {
        fields: null
    },
    TestCaseResult: {
        fields: null
    },
    TestCaseResult2: {
        fields: null
    },
    TestCaseResultAttachmentModel: {
        fields: null
    },
    TestCaseResultIdentifier: {
        fields: null
    },
    TestCaseResultUpdateModel: {
        fields: null
    },
    TestConfiguration: {
        fields: null
    },
    TestConfigurationState: {
        enumValues: {
            "active": 1,
            "inactive": 2,
        }
    },
    TestEnvironment: {
        fields: null
    },
    TestFailureDetails: {
        fields: null
    },
    TestFailuresAnalysis: {
        fields: null
    },
    TestIterationDetailsModel: {
        fields: null
    },
    TestMessageLogDetails: {
        fields: null
    },
    TestOutcome: {
        enumValues: {
            "unspecified": 0,
            "none": 1,
            "passed": 2,
            "failed": 3,
            "inconclusive": 4,
            "timeout": 5,
            "aborted": 6,
            "blocked": 7,
            "notExecuted": 8,
            "warning": 9,
            "error": 10,
            "notApplicable": 11,
            "paused": 12,
            "inProgress": 13,
            "maxValue": 13,
        }
    },
    TestPlan: {
        fields: null
    },
    TestPlanCloneRequest: {
        fields: null
    },
    TestPlansWithSelection: {
        fields: null
    },
    TestPoint: {
        fields: null
    },
    TestResolutionState: {
        fields: null
    },
    TestResultCreateModel: {
        fields: null
    },
    TestResultModelBase: {
        fields: null
    },
    TestResultParameterModel: {
        fields: null
    },
    TestResultsContext: {
        fields: null
    },
    TestResultsContextType: {
        enumValues: {
            "build": 1,
            "release": 2,
        }
    },
    TestResultsDetails: {
        fields: null
    },
    TestResultsDetailsForGroup: {
        fields: null
    },
    TestResultSummary: {
        fields: null
    },
    TestResultTrendFilter: {
        fields: null
    },
    TestRun: {
        fields: null
    },
    TestRunCoverage: {
        fields: null
    },
    TestRunState: {
        enumValues: {
            "unspecified": 0,
            "notStarted": 1,
            "inProgress": 2,
            "completed": 3,
            "aborted": 4,
            "waiting": 5,
            "needsInvestigation": 6,
        }
    },
    TestRunStatistic: {
        fields: null
    },
    TestRunSubstate: {
        enumValues: {
            "none": 0,
            "creatingEnvironment": 1,
            "runningTests": 2,
            "canceledByUser": 3,
            "abortedBySystem": 4,
            "timedOut": 5,
            "pendingAnalysis": 6,
            "analyzed": 7,
            "cancellationInProgress": 8,
        }
    },
    TestSettings: {
        fields: null
    },
    TestSuite: {
        fields: null
    },
    TestSuiteCloneRequest: {
        fields: null
    },
    TestVariable: {
        fields: null
    },
    WorkItemReference: {
        fields: null
    },
};
exports.TypeInfo.AggregatedDataForResultTrend.fields = {
    resultsByOutcome: {},
    testResultsContext: {
        typeInfo: exports.TypeInfo.TestResultsContext
    },
};
exports.TypeInfo.AggregatedResultsAnalysis.fields = {
    previousContext: {
        typeInfo: exports.TypeInfo.TestResultsContext
    },
    resultsByOutcome: {},
    resultsDifference: {
        typeInfo: exports.TypeInfo.AggregatedResultsDifference
    },
};
exports.TypeInfo.AggregatedResultsByOutcome.fields = {
    outcome: {
        enumType: exports.TypeInfo.TestOutcome
    },
};
exports.TypeInfo.AggregatedResultsDifference.fields = {};
exports.TypeInfo.BatchResponse.fields = {
    responses: {
        isArray: true,
        typeInfo: exports.TypeInfo.Response
    },
};
exports.TypeInfo.BuildConfiguration.fields = {
    project: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.BuildCoverage.fields = {
    configuration: {
        typeInfo: exports.TypeInfo.BuildConfiguration
    },
    modules: {
        isArray: true,
        typeInfo: exports.TypeInfo.ModuleCoverage
    },
};
exports.TypeInfo.BuildReference.fields = {};
exports.TypeInfo.CloneOperationInformation.fields = {
    cloneStatistics: {
        typeInfo: exports.TypeInfo.CloneStatistics
    },
    completionDate: {
        isDate: true,
    },
    creationDate: {
        isDate: true,
    },
    destinationObject: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    destinationPlan: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    destinationProject: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    resultObjectType: {
        enumType: exports.TypeInfo.ResultObjectType
    },
    sourceObject: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    sourcePlan: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    sourceProject: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    state: {
        enumType: exports.TypeInfo.CloneOperationState
    },
};
exports.TypeInfo.CloneOptions.fields = {};
exports.TypeInfo.CloneStatistics.fields = {};
exports.TypeInfo.CodeCoverageData.fields = {
    coverageStats: {
        isArray: true,
        typeInfo: exports.TypeInfo.CodeCoverageStatistics
    },
};
exports.TypeInfo.CodeCoverageStatistics.fields = {};
exports.TypeInfo.CodeCoverageSummary.fields = {
    build: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    coverageData: {
        isArray: true,
        typeInfo: exports.TypeInfo.CodeCoverageData
    },
    deltaBuild: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.CoverageStatistics.fields = {};
exports.TypeInfo.CustomTestField.fields = {};
exports.TypeInfo.CustomTestFieldDefinition.fields = {
    fieldType: {
        enumType: exports.TypeInfo.CustomTestFieldType
    },
    scope: {
        enumType: exports.TypeInfo.CustomTestFieldScope
    },
};
exports.TypeInfo.DtlEnvironmentDetails.fields = {};
exports.TypeInfo.FailingSince.fields = {
    build: {
        typeInfo: exports.TypeInfo.BuildReference
    },
    date: {
        isDate: true,
    },
    release: {
        typeInfo: exports.TypeInfo.ReleaseReference
    },
};
exports.TypeInfo.FunctionCoverage.fields = {
    statistics: {
        typeInfo: exports.TypeInfo.CoverageStatistics
    },
};
exports.TypeInfo.LastResultDetails.fields = {
    dateCompleted: {
        isDate: true,
    },
    runBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
};
exports.TypeInfo.ModuleCoverage.fields = {
    functions: {
        isArray: true,
        typeInfo: exports.TypeInfo.FunctionCoverage
    },
    statistics: {
        typeInfo: exports.TypeInfo.CoverageStatistics
    },
};
exports.TypeInfo.NameValuePair.fields = {};
exports.TypeInfo.PlanUpdateModel.fields = {
    area: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    automatedTestEnvironment: {
        typeInfo: exports.TypeInfo.TestEnvironment
    },
    automatedTestSettings: {
        typeInfo: exports.TypeInfo.TestSettings
    },
    build: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    manualTestEnvironment: {
        typeInfo: exports.TypeInfo.TestEnvironment
    },
    manualTestSettings: {
        typeInfo: exports.TypeInfo.TestSettings
    },
    owner: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
};
exports.TypeInfo.PointAssignment.fields = {
    configuration: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    tester: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
};
exports.TypeInfo.PointUpdateModel.fields = {};
exports.TypeInfo.PointWorkItemProperty.fields = {};
exports.TypeInfo.QueryModel.fields = {};
exports.TypeInfo.ReleaseReference.fields = {};
exports.TypeInfo.Response.fields = {};
exports.TypeInfo.ResultRetentionSettings.fields = {
    lastUpdatedBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    lastUpdatedDate: {
        isDate: true,
    },
};
exports.TypeInfo.ResultUpdateRequestModel.fields = {
    actionResultDeletes: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestActionResultModel
    },
    actionResults: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestActionResultModel
    },
    parameterDeletes: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestResultParameterModel
    },
    parameters: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestResultParameterModel
    },
    testCaseResult: {
        typeInfo: exports.TypeInfo.TestCaseResultUpdateModel
    },
};
exports.TypeInfo.ResultUpdateResponseModel.fields = {};
exports.TypeInfo.RunCreateModel.fields = {
    build: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    customTestFields: {
        isArray: true,
        typeInfo: exports.TypeInfo.CustomTestField
    },
    dtlAutEnvironment: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    dtlTestEnvironment: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    environmentDetails: {
        typeInfo: exports.TypeInfo.DtlEnvironmentDetails
    },
    filter: {
        typeInfo: exports.TypeInfo.RunFilter
    },
    owner: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    plan: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    testSettings: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.RunFilter.fields = {};
exports.TypeInfo.RunStatistic.fields = {
    resolutionState: {
        typeInfo: exports.TypeInfo.TestResolutionState
    },
};
exports.TypeInfo.RunUpdateModel.fields = {
    build: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    dtlAutEnvironment: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    dtlEnvironment: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    dtlEnvironmentDetails: {
        typeInfo: exports.TypeInfo.DtlEnvironmentDetails
    },
    logEntries: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestMessageLogDetails
    },
    substate: {
        enumType: exports.TypeInfo.TestRunSubstate
    },
    testSettings: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.ShallowReference.fields = {};
exports.TypeInfo.SharedStepModel.fields = {};
exports.TypeInfo.SuiteCreateModel.fields = {};
exports.TypeInfo.SuiteTestCase.fields = {
    pointAssignments: {
        isArray: true,
        typeInfo: exports.TypeInfo.PointAssignment
    },
    testCase: {
        typeInfo: exports.TypeInfo.WorkItemReference
    },
};
exports.TypeInfo.SuiteUpdateModel.fields = {};
exports.TypeInfo.TestActionResultModel.fields = {
    completedDate: {
        isDate: true,
    },
    sharedStepModel: {
        typeInfo: exports.TypeInfo.SharedStepModel
    },
    startedDate: {
        isDate: true,
    },
};
exports.TypeInfo.TestAttachmentReference.fields = {};
exports.TypeInfo.TestAttachmentRequestModel.fields = {};
exports.TypeInfo.TestCaseResult.fields = {
    area: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    associatedBugs: {
        isArray: true,
        typeInfo: exports.TypeInfo.ShallowReference
    },
    build: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    buildReference: {
        typeInfo: exports.TypeInfo.BuildReference
    },
    completedDate: {
        isDate: true,
    },
    configuration: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    createdDate: {
        isDate: true,
    },
    customFields: {
        isArray: true,
        typeInfo: exports.TypeInfo.CustomTestField
    },
    failingSince: {
        typeInfo: exports.TypeInfo.FailingSince
    },
    iterationDetails: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestIterationDetailsModel
    },
    lastUpdatedBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    lastUpdatedDate: {
        isDate: true,
    },
    owner: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    project: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    releaseReference: {
        typeInfo: exports.TypeInfo.ReleaseReference
    },
    runBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    startedDate: {
        isDate: true,
    },
    testCase: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    testPoint: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    testRun: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.TestCaseResult2.fields = {
    endTime: {
        isDate: true,
    },
    outcome: {
        enumType: exports.TypeInfo.ResultOutcome
    },
    startTime: {
        isDate: true,
    },
};
exports.TypeInfo.TestCaseResultAttachmentModel.fields = {};
exports.TypeInfo.TestCaseResultIdentifier.fields = {};
exports.TypeInfo.TestCaseResultUpdateModel.fields = {
    customFields: {
        isArray: true,
        typeInfo: exports.TypeInfo.CustomTestField
    },
    owner: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    runBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    testResult: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.TestConfiguration.fields = {
    area: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    lastUpdatedBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    lastUpdatedDate: {
        isDate: true,
    },
    project: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    state: {
        enumType: exports.TypeInfo.TestConfigurationState
    },
    values: {
        isArray: true,
        typeInfo: exports.TypeInfo.NameValuePair
    },
};
exports.TypeInfo.TestEnvironment.fields = {};
exports.TypeInfo.TestFailureDetails.fields = {
    testResults: {
        isArray: true,
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.TestFailuresAnalysis.fields = {
    existingFailures: {
        typeInfo: exports.TypeInfo.TestFailureDetails
    },
    fixedTests: {
        typeInfo: exports.TypeInfo.TestFailureDetails
    },
    newFailures: {
        typeInfo: exports.TypeInfo.TestFailureDetails
    },
    previousContext: {
        typeInfo: exports.TypeInfo.TestResultsContext
    },
};
exports.TypeInfo.TestIterationDetailsModel.fields = {
    actionResults: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestActionResultModel
    },
    attachments: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestCaseResultAttachmentModel
    },
    completedDate: {
        isDate: true,
    },
    parameters: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestResultParameterModel
    },
    startedDate: {
        isDate: true,
    },
};
exports.TypeInfo.TestMessageLogDetails.fields = {
    dateCreated: {
        isDate: true,
    },
};
exports.TypeInfo.TestPlan.fields = {
    area: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    automatedTestEnvironment: {
        typeInfo: exports.TypeInfo.TestEnvironment
    },
    automatedTestSettings: {
        typeInfo: exports.TypeInfo.TestSettings
    },
    build: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    endDate: {
        isDate: true,
    },
    manualTestEnvironment: {
        typeInfo: exports.TypeInfo.TestEnvironment
    },
    manualTestSettings: {
        typeInfo: exports.TypeInfo.TestSettings
    },
    owner: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    previousBuild: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    project: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    rootSuite: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    startDate: {
        isDate: true,
    },
    updatedBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    updatedDate: {
        isDate: true,
    },
};
exports.TypeInfo.TestPlanCloneRequest.fields = {
    cloneOptions: {
        typeInfo: exports.TypeInfo.CloneOptions
    },
    destinationTestPlan: {
        typeInfo: exports.TypeInfo.TestPlan
    },
};
exports.TypeInfo.TestPlansWithSelection.fields = {
    plans: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestPlan
    },
};
exports.TypeInfo.TestPoint.fields = {
    assignedTo: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    configuration: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    lastResult: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    lastResultDetails: {
        typeInfo: exports.TypeInfo.LastResultDetails
    },
    lastTestRun: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    lastUpdatedBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    lastUpdatedDate: {
        isDate: true,
    },
    suite: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    testCase: {
        typeInfo: exports.TypeInfo.WorkItemReference
    },
    testPlan: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.TestResolutionState.fields = {
    project: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.TestResultCreateModel.fields = {
    area: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    configuration: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    customFields: {
        isArray: true,
        typeInfo: exports.TypeInfo.CustomTestField
    },
    owner: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    runBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    testCase: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    testPoint: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.TestResultModelBase.fields = {
    completedDate: {
        isDate: true,
    },
    startedDate: {
        isDate: true,
    },
};
exports.TypeInfo.TestResultParameterModel.fields = {};
exports.TypeInfo.TestResultsContext.fields = {
    build: {
        typeInfo: exports.TypeInfo.BuildReference
    },
    contextType: {
        enumType: exports.TypeInfo.TestResultsContextType
    },
    release: {
        typeInfo: exports.TypeInfo.ReleaseReference
    },
};
exports.TypeInfo.TestResultsDetails.fields = {
    resultsForGroup: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestResultsDetailsForGroup
    },
};
exports.TypeInfo.TestResultsDetailsForGroup.fields = {
    ids: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestCaseResultIdentifier
    },
    resultsCountByOutcome: {},
};
exports.TypeInfo.TestResultSummary.fields = {
    aggregatedResultsAnalysis: {
        typeInfo: exports.TypeInfo.AggregatedResultsAnalysis
    },
    teamProject: {
        typeInfo: TfsCoreInterfaces.TypeInfo.TeamProjectReference
    },
    testFailures: {
        typeInfo: exports.TypeInfo.TestFailuresAnalysis
    },
    testResultsContext: {
        typeInfo: exports.TypeInfo.TestResultsContext
    },
};
exports.TypeInfo.TestResultTrendFilter.fields = {};
exports.TypeInfo.TestRun.fields = {
    build: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    buildConfiguration: {
        typeInfo: exports.TypeInfo.BuildConfiguration
    },
    completedDate: {
        isDate: true,
    },
    createdDate: {
        isDate: true,
    },
    customFields: {
        isArray: true,
        typeInfo: exports.TypeInfo.CustomTestField
    },
    dtlAutEnvironment: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    dtlEnvironment: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    dtlEnvironmentCreationDetails: {
        typeInfo: exports.TypeInfo.DtlEnvironmentDetails
    },
    dueDate: {
        isDate: true,
    },
    filter: {
        typeInfo: exports.TypeInfo.RunFilter
    },
    lastUpdatedBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    lastUpdatedDate: {
        isDate: true,
    },
    owner: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    plan: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    project: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    runStatistics: {
        isArray: true,
        typeInfo: exports.TypeInfo.RunStatistic
    },
    startedDate: {
        isDate: true,
    },
    substate: {
        enumType: exports.TypeInfo.TestRunSubstate
    },
    testEnvironment: {
        typeInfo: exports.TypeInfo.TestEnvironment
    },
    testSettings: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.TestRunCoverage.fields = {
    modules: {
        isArray: true,
        typeInfo: exports.TypeInfo.ModuleCoverage
    },
    testRun: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.TestRunStatistic.fields = {
    run: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    runStatistics: {
        isArray: true,
        typeInfo: exports.TypeInfo.RunStatistic
    },
};
exports.TypeInfo.TestSettings.fields = {};
exports.TypeInfo.TestSuite.fields = {
    children: {
        isArray: true,
        typeInfo: exports.TypeInfo.TestSuite
    },
    defaultConfigurations: {
        isArray: true,
        typeInfo: exports.TypeInfo.ShallowReference
    },
    lastPopulatedDate: {
        isDate: true,
    },
    lastUpdatedBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    lastUpdatedDate: {
        isDate: true,
    },
    parent: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    plan: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    project: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
    suites: {
        isArray: true,
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.TestSuiteCloneRequest.fields = {
    cloneOptions: {
        typeInfo: exports.TypeInfo.CloneOptions
    },
};
exports.TypeInfo.TestVariable.fields = {
    project: {
        typeInfo: exports.TypeInfo.ShallowReference
    },
};
exports.TypeInfo.WorkItemReference.fields = {};
