// vso-node-api/CoreApi.js
/*
* ---------------------------------------------------------
* Copyright(C) Microsoft Corporation. All rights reserved.
* ---------------------------------------------------------
*
* ---------------------------------------------------------
* Generated file, DO NOT EDIT
* ---------------------------------------------------------
*/
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
// Licensed under the MIT license.  See LICENSE file in the project root for full license information.
/// <reference path="../node/node.d.ts"/>
/// <reference path="../q/Q.d.ts"/>
var Q = require('q');
var basem = require('./ClientApiBases');
var CoreInterfaces = require("./interfaces/CoreInterfaces");
var OperationsInterfaces = require("./interfaces/common/OperationsInterfaces");
var VSSInterfaces = require("./interfaces/common/VSSInterfaces");
var CoreApi = (function (_super) {
    __extends(CoreApi, _super);
    function CoreApi(baseUrl, handlers) {
        _super.call(this, baseUrl, handlers, 'node-Core-api');
    }
    /**
     * @param {CoreInterfaces.WebApiConnectedServiceDetails} connectedServiceCreationData
     * @param {string} projectId
     * @param onResult callback function with the resulting CoreInterfaces.WebApiConnectedService
     */
    CoreApi.prototype.createConnectedService = function (connectedServiceCreationData, projectId, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "b4f70219-e18b-42c5-abe3-98b07d35525e", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: CoreInterfaces.TypeInfo.WebApiConnectedServiceDetails, responseTypeMetadata: CoreInterfaces.TypeInfo.WebApiConnectedService, responseIsCollection: false };
            _this.restClient.create(url, apiVersion, connectedServiceCreationData, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * @param {string} projectId
     * @param {string} name
     * @param onResult callback function with the resulting CoreInterfaces.WebApiConnectedServiceDetails
     */
    CoreApi.prototype.getConnectedServiceDetails = function (projectId, name, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId,
            name: name
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "b4f70219-e18b-42c5-abe3-98b07d35525e", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.WebApiConnectedServiceDetails, responseIsCollection: false };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * @param {string} projectId
     * @param {CoreInterfaces.ConnectedServiceKind} kind
     * @param onResult callback function with the resulting CoreInterfaces.WebApiConnectedService[]
     */
    CoreApi.prototype.getConnectedServices = function (projectId, kind, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId
        };
        var queryValues = {
            kind: kind,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "b4f70219-e18b-42c5-abe3-98b07d35525e", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.WebApiConnectedService, responseIsCollection: true };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * @param {CoreInterfaces.IdentityData} mruData
     * @param {string} mruName
     * @param onResult callback function
     */
    CoreApi.prototype.createIdentityMru = function (mruData, mruName, onResult) {
        var _this = this;
        var routeValues = {
            mruName: mruName
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "5ead0b70-2572-4697-97e9-f341069a783a", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: CoreInterfaces.TypeInfo.IdentityData, responseIsCollection: false };
            _this.restClient.create(url, apiVersion, mruData, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
    };
    /**
     * @param {CoreInterfaces.IdentityData} mruData
     * @param {string} mruName
     * @param onResult callback function
     */
    CoreApi.prototype.deleteIdentityMru = function (mruData, mruName, onResult) {
        var _this = this;
        var routeValues = {
            mruName: mruName
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "5ead0b70-2572-4697-97e9-f341069a783a", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: CoreInterfaces.TypeInfo.IdentityData, responseIsCollection: false };
            _this.restClient.delete(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
    };
    /**
     * @param {string} mruName
     * @param onResult callback function with the resulting VSSInterfaces.IdentityRef[]
     */
    CoreApi.prototype.getIdentityMru = function (mruName, onResult) {
        var _this = this;
        var routeValues = {
            mruName: mruName
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "5ead0b70-2572-4697-97e9-f341069a783a", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: VSSInterfaces.TypeInfo.IdentityRef, responseIsCollection: true };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * @param {CoreInterfaces.IdentityData} mruData
     * @param {string} mruName
     * @param onResult callback function
     */
    CoreApi.prototype.updateIdentityMru = function (mruData, mruName, onResult) {
        var _this = this;
        var routeValues = {
            mruName: mruName
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "5ead0b70-2572-4697-97e9-f341069a783a", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: CoreInterfaces.TypeInfo.IdentityData, responseIsCollection: false };
            _this.restClient.update(url, apiVersion, mruData, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
    };
    /**
     * @param {string} projectId
     * @param {string} teamId
     * @param {number} top
     * @param {number} skip
     * @param onResult callback function with the resulting VSSInterfaces.IdentityRef[]
     */
    CoreApi.prototype.getTeamMembers = function (projectId, teamId, top, skip, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId,
            teamId: teamId
        };
        var queryValues = {
            '$top': top,
            '$skip': skip,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "294c494c-2600-4d7e-b76c-3dd50c3c95be", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: VSSInterfaces.TypeInfo.IdentityRef, responseIsCollection: true };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * Retrieve process by id
     *
     * @param {string} processId
     * @param onResult callback function with the resulting CoreInterfaces.Process
     */
    CoreApi.prototype.getProcessById = function (processId, onResult) {
        var _this = this;
        var routeValues = {
            processId: processId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "93878975-88c5-4e6a-8abb-7ddd77a8a7d8", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.Process, responseIsCollection: false };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * @param onResult callback function with the resulting CoreInterfaces.Process[]
     */
    CoreApi.prototype.getProcesses = function (onResult) {
        var _this = this;
        var routeValues = {};
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "93878975-88c5-4e6a-8abb-7ddd77a8a7d8", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.Process, responseIsCollection: true };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * Get project collection with the specified id or name.
     *
     * @param {string} collectionId
     * @param onResult callback function with the resulting CoreInterfaces.TeamProjectCollection
     */
    CoreApi.prototype.getProjectCollection = function (collectionId, onResult) {
        var _this = this;
        var routeValues = {
            collectionId: collectionId
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "core", "8031090f-ef1d-4af6-85fc-698cd75d42bf", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.TeamProjectCollection, responseIsCollection: false };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * Get project collection references for this application.
     *
     * @param {number} top
     * @param {number} skip
     * @param onResult callback function with the resulting CoreInterfaces.TeamProjectCollectionReference[]
     */
    CoreApi.prototype.getProjectCollections = function (top, skip, onResult) {
        var _this = this;
        var routeValues = {};
        var queryValues = {
            '$top': top,
            '$skip': skip,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "core", "8031090f-ef1d-4af6-85fc-698cd75d42bf", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.TeamProjectCollectionReference, responseIsCollection: true };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * @param {number} minRevision
     * @param onResult callback function with the resulting CoreInterfaces.TeamProjectReference[]
     */
    CoreApi.prototype.getProjectHistory = function (minRevision, onResult) {
        var _this = this;
        var routeValues = {};
        var queryValues = {
            minRevision: minRevision,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "6488a877-4749-4954-82ea-7340d36be9f2", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.TeamProjectReference, responseIsCollection: true };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * Get project with the specified id or name, optionally including capabilities.
     *
     * @param {string} projectId
     * @param {boolean} includeCapabilities - Include capabilities (such as source control) in the team project result (default: false).
     * @param {boolean} includeHistory - Search within renamed projects (that had such name in the past).
     * @param onResult callback function with the resulting CoreInterfaces.TeamProject
     */
    CoreApi.prototype.getProject = function (projectId, includeCapabilities, includeHistory, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId
        };
        var queryValues = {
            includeCapabilities: includeCapabilities,
            includeHistory: includeHistory,
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "core", "603fe2ac-9723-48b9-88ad-09305aa6c6e1", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.TeamProject, responseIsCollection: false };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * Get project references with the specified state
     *
     * @param {any} stateFilter - Filter on team projects in a specific team project state (default: WellFormed).
     * @param {number} top
     * @param {number} skip
     * @param onResult callback function with the resulting CoreInterfaces.TeamProjectReference[]
     */
    CoreApi.prototype.getProjects = function (stateFilter, top, skip, onResult) {
        var _this = this;
        var routeValues = {};
        var queryValues = {
            stateFilter: stateFilter,
            '$top': top,
            '$skip': skip,
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "core", "603fe2ac-9723-48b9-88ad-09305aa6c6e1", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.TeamProjectReference, responseIsCollection: true };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * Queue a project creation.
     *
     * @param {CoreInterfaces.TeamProject} projectToCreate - The project to create.
     * @param onResult callback function with the resulting OperationsInterfaces.OperationReference
     */
    CoreApi.prototype.queueCreateProject = function (projectToCreate, onResult) {
        var _this = this;
        var routeValues = {};
        this.vsoClient.getVersioningData("3.0-preview.3", "core", "603fe2ac-9723-48b9-88ad-09305aa6c6e1", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: CoreInterfaces.TypeInfo.TeamProject, responseTypeMetadata: OperationsInterfaces.TypeInfo.OperationReference, responseIsCollection: false };
            _this.restClient.create(url, apiVersion, projectToCreate, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * Queue a project deletion.
     *
     * @param {string} projectId - The project id of the project to delete.
     * @param onResult callback function with the resulting OperationsInterfaces.OperationReference
     */
    CoreApi.prototype.queueDeleteProject = function (projectId, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "core", "603fe2ac-9723-48b9-88ad-09305aa6c6e1", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: OperationsInterfaces.TypeInfo.OperationReference, responseIsCollection: false };
            _this.restClient.delete(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * Update an existing project's name, abbreviation, or description.
     *
     * @param {CoreInterfaces.TeamProject} projectUpdate - The updates for the project.
     * @param {string} projectId - The project id of the project to update.
     * @param onResult callback function with the resulting OperationsInterfaces.OperationReference
     */
    CoreApi.prototype.updateProject = function (projectUpdate, projectId, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "core", "603fe2ac-9723-48b9-88ad-09305aa6c6e1", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: CoreInterfaces.TypeInfo.TeamProject, responseTypeMetadata: OperationsInterfaces.TypeInfo.OperationReference, responseIsCollection: false };
            _this.restClient.update(url, apiVersion, projectUpdate, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * @param {string} proxyUrl
     * @param onResult callback function with the resulting CoreInterfaces.Proxy[]
     */
    CoreApi.prototype.getProxies = function (proxyUrl, onResult) {
        var _this = this;
        var routeValues = {};
        var queryValues = {
            proxyUrl: proxyUrl,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "ec1f4311-f2b4-4c15-b2b8-8990b80d2908", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.Proxy, responseIsCollection: true };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * Creates a team
     *
     * @param {CoreInterfaces.WebApiTeam} team - The team data used to create the team.
     * @param {string} projectId - The name or id (GUID) of the team project in which to create the team.
     * @param onResult callback function with the resulting CoreInterfaces.WebApiTeam
     */
    CoreApi.prototype.createTeam = function (team, projectId, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "d30a3dd1-f8ba-442a-b86a-bd0c0c383e59", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: CoreInterfaces.TypeInfo.WebApiTeam, responseTypeMetadata: CoreInterfaces.TypeInfo.WebApiTeam, responseIsCollection: false };
            _this.restClient.create(url, apiVersion, team, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * Deletes a team
     *
     * @param {string} projectId - The name or id (GUID) of the team project containing the team to delete.
     * @param {string} teamId - The name of id of the team to delete.
     * @param onResult callback function
     */
    CoreApi.prototype.deleteTeam = function (projectId, teamId, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId,
            teamId: teamId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "d30a3dd1-f8ba-442a-b86a-bd0c0c383e59", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.delete(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
    };
    /**
     * Gets a team
     *
     * @param {string} projectId
     * @param {string} teamId
     * @param onResult callback function with the resulting CoreInterfaces.WebApiTeam
     */
    CoreApi.prototype.getTeam = function (projectId, teamId, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId,
            teamId: teamId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "d30a3dd1-f8ba-442a-b86a-bd0c0c383e59", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.WebApiTeam, responseIsCollection: false };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * @param {string} projectId
     * @param {number} top
     * @param {number} skip
     * @param onResult callback function with the resulting CoreInterfaces.WebApiTeam[]
     */
    CoreApi.prototype.getTeams = function (projectId, top, skip, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId
        };
        var queryValues = {
            '$top': top,
            '$skip': skip,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "d30a3dd1-f8ba-442a-b86a-bd0c0c383e59", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: CoreInterfaces.TypeInfo.WebApiTeam, responseIsCollection: true };
            _this.restClient.getJson(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    /**
     * Updates a team's name and/or description
     *
     * @param {CoreInterfaces.WebApiTeam} teamData
     * @param {string} projectId - The name or id (GUID) of the team project containing the team to update.
     * @param {string} teamId - The name of id of the team to update.
     * @param onResult callback function with the resulting CoreInterfaces.WebApiTeam
     */
    CoreApi.prototype.updateTeam = function (teamData, projectId, teamId, onResult) {
        var _this = this;
        var routeValues = {
            projectId: projectId,
            teamId: teamId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "core", "d30a3dd1-f8ba-442a-b86a-bd0c0c383e59", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: CoreInterfaces.TypeInfo.WebApiTeam, responseTypeMetadata: CoreInterfaces.TypeInfo.WebApiTeam, responseIsCollection: false };
            _this.restClient.update(url, apiVersion, teamData, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
    };
    return CoreApi;
}(basem.ClientApiBase));
exports.CoreApi = CoreApi;
var QCoreApi = (function (_super) {
    __extends(QCoreApi, _super);
    function QCoreApi(baseUrl, handlers) {
        _super.call(this, baseUrl, handlers, CoreApi);
    }
    /**
    * @param {CoreInterfaces.WebApiConnectedServiceDetails} connectedServiceCreationData
    * @param {string} projectId
    */
    QCoreApi.prototype.createConnectedService = function (connectedServiceCreationData, projectId) {
        var deferred = Q.defer();
        this.api.createConnectedService(connectedServiceCreationData, projectId, function (err, statusCode, connectedService) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(connectedService);
            }
        });
        return deferred.promise;
    };
    /**
    * @param {string} projectId
    * @param {string} name
    */
    QCoreApi.prototype.getConnectedServiceDetails = function (projectId, name) {
        var deferred = Q.defer();
        this.api.getConnectedServiceDetails(projectId, name, function (err, statusCode, connectedService) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(connectedService);
            }
        });
        return deferred.promise;
    };
    /**
    * @param {string} projectId
    * @param {CoreInterfaces.ConnectedServiceKind} kind
    */
    QCoreApi.prototype.getConnectedServices = function (projectId, kind) {
        var deferred = Q.defer();
        this.api.getConnectedServices(projectId, kind, function (err, statusCode, connectedServices) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(connectedServices);
            }
        });
        return deferred.promise;
    };
    /**
    * @param {CoreInterfaces.IdentityData} mruData
    * @param {string} mruName
    */
    QCoreApi.prototype.createIdentityMru = function (mruData, mruName) {
        var deferred = Q.defer();
        this.api.createIdentityMru(mruData, mruName, function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        });
        return deferred.promise;
    };
    /**
    * @param {CoreInterfaces.IdentityData} mruData
    * @param {string} mruName
    */
    QCoreApi.prototype.deleteIdentityMru = function (mruData, mruName) {
        var deferred = Q.defer();
        this.api.deleteIdentityMru(mruData, mruName, function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        });
        return deferred.promise;
    };
    /**
    * @param {string} mruName
    */
    QCoreApi.prototype.getIdentityMru = function (mruName) {
        var deferred = Q.defer();
        this.api.getIdentityMru(mruName, function (err, statusCode, identityMru) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(identityMru);
            }
        });
        return deferred.promise;
    };
    /**
    * @param {CoreInterfaces.IdentityData} mruData
    * @param {string} mruName
    */
    QCoreApi.prototype.updateIdentityMru = function (mruData, mruName) {
        var deferred = Q.defer();
        this.api.updateIdentityMru(mruData, mruName, function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        });
        return deferred.promise;
    };
    /**
    * @param {string} projectId
    * @param {string} teamId
    * @param {number} top
    * @param {number} skip
    */
    QCoreApi.prototype.getTeamMembers = function (projectId, teamId, top, skip) {
        var deferred = Q.defer();
        this.api.getTeamMembers(projectId, teamId, top, skip, function (err, statusCode, members) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(members);
            }
        });
        return deferred.promise;
    };
    /**
    * Retrieve process by id
    *
    * @param {string} processId
    */
    QCoreApi.prototype.getProcessById = function (processId) {
        var deferred = Q.defer();
        this.api.getProcessById(processId, function (err, statusCode, processe) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(processe);
            }
        });
        return deferred.promise;
    };
    /**
    */
    QCoreApi.prototype.getProcesses = function () {
        var deferred = Q.defer();
        this.api.getProcesses(function (err, statusCode, processes) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(processes);
            }
        });
        return deferred.promise;
    };
    /**
    * Get project collection with the specified id or name.
    *
    * @param {string} collectionId
    */
    QCoreApi.prototype.getProjectCollection = function (collectionId) {
        var deferred = Q.defer();
        this.api.getProjectCollection(collectionId, function (err, statusCode, projectCollection) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(projectCollection);
            }
        });
        return deferred.promise;
    };
    /**
    * Get project collection references for this application.
    *
    * @param {number} top
    * @param {number} skip
    */
    QCoreApi.prototype.getProjectCollections = function (top, skip) {
        var deferred = Q.defer();
        this.api.getProjectCollections(top, skip, function (err, statusCode, projectCollections) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(projectCollections);
            }
        });
        return deferred.promise;
    };
    /**
    * @param {number} minRevision
    */
    QCoreApi.prototype.getProjectHistory = function (minRevision) {
        var deferred = Q.defer();
        this.api.getProjectHistory(minRevision, function (err, statusCode, projectHistory) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(projectHistory);
            }
        });
        return deferred.promise;
    };
    /**
    * Get project with the specified id or name, optionally including capabilities.
    *
    * @param {string} projectId
    * @param {boolean} includeCapabilities - Include capabilities (such as source control) in the team project result (default: false).
    * @param {boolean} includeHistory - Search within renamed projects (that had such name in the past).
    */
    QCoreApi.prototype.getProject = function (projectId, includeCapabilities, includeHistory) {
        var deferred = Q.defer();
        this.api.getProject(projectId, includeCapabilities, includeHistory, function (err, statusCode, project) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(project);
            }
        });
        return deferred.promise;
    };
    /**
    * Get project references with the specified state
    *
    * @param {any} stateFilter - Filter on team projects in a specific team project state (default: WellFormed).
    * @param {number} top
    * @param {number} skip
    */
    QCoreApi.prototype.getProjects = function (stateFilter, top, skip) {
        var deferred = Q.defer();
        this.api.getProjects(stateFilter, top, skip, function (err, statusCode, projects) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(projects);
            }
        });
        return deferred.promise;
    };
    /**
    * Queue a project creation.
    *
    * @param {CoreInterfaces.TeamProject} projectToCreate - The project to create.
    */
    QCoreApi.prototype.queueCreateProject = function (projectToCreate) {
        var deferred = Q.defer();
        this.api.queueCreateProject(projectToCreate, function (err, statusCode, project) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(project);
            }
        });
        return deferred.promise;
    };
    /**
    * Queue a project deletion.
    *
    * @param {string} projectId - The project id of the project to delete.
    */
    QCoreApi.prototype.queueDeleteProject = function (projectId) {
        var deferred = Q.defer();
        this.api.queueDeleteProject(projectId, function (err, statusCode, project) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(project);
            }
        });
        return deferred.promise;
    };
    /**
    * Update an existing project's name, abbreviation, or description.
    *
    * @param {CoreInterfaces.TeamProject} projectUpdate - The updates for the project.
    * @param {string} projectId - The project id of the project to update.
    */
    QCoreApi.prototype.updateProject = function (projectUpdate, projectId) {
        var deferred = Q.defer();
        this.api.updateProject(projectUpdate, projectId, function (err, statusCode, project) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(project);
            }
        });
        return deferred.promise;
    };
    /**
    * @param {string} proxyUrl
    */
    QCoreApi.prototype.getProxies = function (proxyUrl) {
        var deferred = Q.defer();
        this.api.getProxies(proxyUrl, function (err, statusCode, proxies) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(proxies);
            }
        });
        return deferred.promise;
    };
    /**
    * Creates a team
    *
    * @param {CoreInterfaces.WebApiTeam} team - The team data used to create the team.
    * @param {string} projectId - The name or id (GUID) of the team project in which to create the team.
    */
    QCoreApi.prototype.createTeam = function (team, projectId) {
        var deferred = Q.defer();
        this.api.createTeam(team, projectId, function (err, statusCode, team) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(team);
            }
        });
        return deferred.promise;
    };
    /**
    * Deletes a team
    *
    * @param {string} projectId - The name or id (GUID) of the team project containing the team to delete.
    * @param {string} teamId - The name of id of the team to delete.
    */
    QCoreApi.prototype.deleteTeam = function (projectId, teamId) {
        var deferred = Q.defer();
        this.api.deleteTeam(projectId, teamId, function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        });
        return deferred.promise;
    };
    /**
    * Gets a team
    *
    * @param {string} projectId
    * @param {string} teamId
    */
    QCoreApi.prototype.getTeam = function (projectId, teamId) {
        var deferred = Q.defer();
        this.api.getTeam(projectId, teamId, function (err, statusCode, team) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(team);
            }
        });
        return deferred.promise;
    };
    /**
    * @param {string} projectId
    * @param {number} top
    * @param {number} skip
    */
    QCoreApi.prototype.getTeams = function (projectId, top, skip) {
        var deferred = Q.defer();
        this.api.getTeams(projectId, top, skip, function (err, statusCode, teams) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(teams);
            }
        });
        return deferred.promise;
    };
    /**
    * Updates a team's name and/or description
    *
    * @param {CoreInterfaces.WebApiTeam} teamData
    * @param {string} projectId - The name or id (GUID) of the team project containing the team to update.
    * @param {string} teamId - The name of id of the team to update.
    */
    QCoreApi.prototype.updateTeam = function (teamData, projectId, teamId) {
        var deferred = Q.defer();
        this.api.updateTeam(teamData, projectId, teamId, function (err, statusCode, team) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(team);
            }
        });
        return deferred.promise;
    };
    return QCoreApi;
}(basem.QClientApiBase));
exports.QCoreApi = QCoreApi;
